File: class181_AdvancedAlgorithmsAndDataStructures/LeetCode_435_NonOverlappingIntervals.java
package class008_AdvancedAlgorithmsAndDataStructures.sweep_line_problems;

import java.util.*;

/**
 * LeetCode 435. Non-overlapping Intervals
 * 
 * 题目描述：
 * 给定一个区间的集合 intervals ，其中 intervals[i] = [start_i, end_i] 。
 * 返回需要移除区间的最小数量，使剩余区间互不重叠。
 * 
 * 解题思路：
 * 这是一个经典的贪心算法问题，也可以用扫描线算法的思想来解决。
 * 我们按照区间的结束时间进行排序，然后贪心地选择结束时间最早的区间，
 * 这样可以为后续的区间留出更多空间。
 * 
 * 时间复杂度：O(n log n)
 * 空间复杂度：O(1)
 */
public class LeetCode_435_NonOverlappingIntervals {
    
    static class Solution {
        public int eraseOverlapIntervals(int[][] intervals) {
            if (intervals == null || intervals.length == 0) {
                return 0;
            }
            
            // 按照区间结束时间排序
            Arrays.sort(intervals, (a, b) -> Integer.compare(a[1], b[1]));
            
            int count = 0; // 需要移除的区间数量
            int end = intervals[0][1]; // 当前选择区间的结束时间
            
            // 遍历所有区间
            for (int i = 1; i < intervals.length; i++) {
                // 如果当前区间的开始时间小于前一个区间的结束时间，说明有重叠
                if (intervals[i][0] < end) {
                    count++; // 移除当前区间
                } else {
                    // 更新结束时间
                    end = intervals[i][1];
                }
            }
            
            return count;
        }
        
        // 另一种解法：计算最多能保留的区间数量
        public int eraseOverlapIntervals2(int[][] intervals) {
            if (intervals == null || intervals.length == 0) {
                return 0;
            }
            
            // 按照区间结束时间排序
            Arrays.sort(intervals, (a, b) -> Integer.compare(a[1], b[1]));
            
            int count = 1; // 能保留的区间数量，至少能保留第一个区间
            int end = intervals[0][1]; // 当前保留区间的结束时间
            
            // 遍历所有区间
            for (int i = 1; i < intervals.length; i++) {
                // 如果当前区间的开始时间大于等于前一个保留区间的结束时间，说明不重叠
                if (intervals[i][0] >= end) {
                    count++; // 保留当前区间
                    end = intervals[i][1]; // 更新结束时间
                }
            }
            
            return intervals.length - count; // 需要移除的数量 = 总数 - 保留的数量
        }
    }
    
    // 测试方法
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // 测试用例1
        int[][] intervals1 = {{1,2},{2,3},{3,4},{1,3}};
        System.out.println("测试用例1:");
        System.out.println("区间: " + Arrays.deepToString(intervals1));
        System.out.println("需要移除的最小区间数: " + solution.eraseOverlapIntervals(intervals1));
        System.out.println("另一种解法结果: " + solution.eraseOverlapIntervals2(intervals1));
        System.out.println();
        
        // 测试用例2
        int[][] intervals2 = {{1,2},{1,2},{1,2}};
        System.out.println("测试用例2:");
        System.out.println("区间: " + Arrays.deepToString(intervals2));
        System.out.println("需要移除的最小区间数: " + solution.eraseOverlapIntervals(intervals2));
        System.out.println("另一种解法结果: " + solution.eraseOverlapIntervals2(intervals2));
        System.out.println();
        
        // 测试用例3
        int[][] intervals3 = {{1,2},{2,3}};
        System.out.println("测试用例3:");
        System.out.println("区间: " + Arrays.deepToString(intervals3));
        System.out.println("需要移除的最小区间数: " + solution.eraseOverlapIntervals(intervals3));
        System.out.println("另一种解法结果: " + solution.eraseOverlapIntervals2(intervals3));
        System.out.println();
        
        // 测试用例4
        int[][] intervals4 = {{1,100},{11,22},{1,11},{2,12}};
        System.out.println("测试用例4:");
        System.out.println("区间: " + Arrays.deepToString(intervals4));
        System.out.println("需要移除的最小区间数: " + solution.eraseOverlapIntervals(intervals4));
        System.out.println("另一种解法结果: " + solution.eraseOverlapIntervals2(intervals4));
    }
}

File: class132_IntervalSchedulingAndSlidingWindow/LeetCode435_NonOverlappingIntervals.cpp
/**
 * LeetCode 435. Non-overlapping Intervals
 * 
 * 题目描述：
 * 给定一个区间的集合 intervals，其中 intervals[i] = [start_i, end_i]。
 * 返回需要移除区间的最小数量，使剩余区间互不重叠。
 * 
 * 解题思路：
 * 这是一个经典的贪心算法问题。为了移除最少的区间，我们应该保留尽可能多的不重叠区间。
 * 
 * 算法步骤：
 * 1. 将所有区间按结束时间排序
 * 2. 使用贪心策略：总是选择结束时间最早的区间
 * 3. 遍历排序后的区间，统计可以保留的区间数量
 * 4. 返回总区间数减去保留的区间数，即为需要移除的区间数
 * 
 * 贪心策略的正确性：
 * 选择结束时间最早的区间可以为后续区间留下更多空间，从而最大化保留的区间数量。
 * 
 * 时间复杂度：O(n * log n)
 * 空间复杂度：O(1)
 * 
 * 相关题目：
 * - LeetCode 1353. 最多可以参加的会议数目 (贪心)
 * - LeetCode 646. 最长数对链 (贪心)
 * - LeetCode 1235. 最大盈利的工作调度 (动态规划 + 二分查找)
 */

// 简化版C++实现，避免使用STL容器
// 由于编译环境限制，使用基本数组和手动实现算法

const int MAX_N = 100005;

// 区间结构体
struct Interval {
    int start, end;
};

Interval intervals[MAX_N];
int n;

// 比较函数，用于按结束时间排序
bool compareIntervals(Interval a, Interval b) {
    return a.end < b.end;
}

// 简单的排序实现（冒泡排序）
void sortIntervals() {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - 1 - i; j++) {
            if (intervals[j].end > intervals[j + 1].end) {
                // 交换
                Interval temp = intervals[j];
                intervals[j] = intervals[j + 1];
                intervals[j + 1] = temp;
            }
        }
    }
}

/**
 * 计算需要移除的最小区间数
 * 
 * @param intervals_input 区间数组
 * @param size 数组大小
 * @return 需要移除的区间数
 */
int eraseOverlapIntervals(int intervals_input[][2], int size) {
    // 边界情况处理
    if (size == 0) {
        return 0;
    }
    
    n = size;
    
    // 将输入转换为内部结构
    for (int i = 0; i < n; i++) {
        intervals[i].start = intervals_input[i][0];
        intervals[i].end = intervals_input[i][1];
    }
    
    // 按结束时间排序
    sortIntervals();
    
    // 初始化计数器和上一个保留区间的结束时间
    int count = 1;  // 至少可以保留一个区间
    int end = intervals[0].end;  // 第一个区间的结束时间
    
    // 遍历剩余区间
    for (int i = 1; i < n; i++) {
        // 如果当前区间的开始时间 >= 上一个保留区间的结束时间
        // 说明不重叠，可以保留
        if (intervals[i].start >= end) {
            count++;
            end = intervals[i].end;  // 更新结束时间
        }
        // 如果重叠，则跳过当前区间（相当于移除）
    }
    
    // 返回需要移除的区间数
    return n - count;
}

// 简单的测试函数
void runTests() {
    // 测试用例1
    int intervals1[][2] = {{1,2},{2,3},{3,4},{1,3}};
    int size1 = 4;
    // 期望输出: 1
    
    // 测试用例2
    int intervals2[][2] = {{1,2},{1,2},{1,2}};
    int size2 = 3;
    // 期望输出: 2
    
    // 测试用例3
    int intervals3[][2] = {{1,2},{2,3}};
    int size3 = 2;
    // 期望输出: 0
}